sql/core/src/test/scala/org/apache/spark/sql/UnwrapCastInComparisonEndToEndSuite.scala

```diff
@@ -0,0 +1,165 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.catalyst.expressions.IntegralLiteralTestUtils.{negativeInt, positiveInt}
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.Decimal
+
+class UnwrapCastInComparisonEndToEndSuite extends QueryTest with SharedSparkSession {
+  import testImplicits._
+
+  test("cases when literal is max") {
+    val t = "test_table"
+    withTable(t) {
+      Seq[(Integer, java.lang.Short, java.lang.Float)](
+        (1, 100.toShort, 3.14.toFloat), (2, Short.MaxValue, Float.NaN), (3, null, null))
+        .toDF("c1", "c2", "c3").write.saveAsTable(t)
+      val df = spark.table(t)
+
+      var lit = Short.MaxValue.toInt
+      checkAnswer(df.select("c1").where(s"c2 > $lit"), Seq.empty)
+      checkAnswer(df.select("c1").where(s"c2 >= $lit"), Row(2))
+      checkAnswer(df.select("c1").where(s"c2 == $lit"), Row(2))
+      checkAnswer(df.select("c1").where(s"c2 <=> $lit"), Row(2))
+      checkAnswer(df.select("c1").where(s"c2 != $lit"), Row(1))
+      checkAnswer(df.select("c1").where(s"c2 <= $lit"), Row(1) :: Row(2) :: Nil)
+      checkAnswer(df.select("c1").where(s"c2 < $lit"), Row(1))
+
+      checkAnswer(df.select("c1").where(s"c3 > double('nan')"), Seq.empty)
+      checkAnswer(df.select("c1").where(s"c3 >= double('nan')"), Row(2))
+      checkAnswer(df.select("c1").where(s"c3 == double('nan')"), Row(2))
+      checkAnswer(df.select("c1").where(s"c3 <=> double('nan')"), Row(2))
+      checkAnswer(df.select("c1").where(s"c3 != double('nan')"), Row(1))
+      checkAnswer(df.select("c1").where(s"c3 <= double('nan')"), Row(1) :: Row(2) :: Nil)
+      checkAnswer(df.select("c1").where(s"c3 < double('nan')"), Row(1))
+
+      lit = positiveInt
+      checkAnswer(df.select("c1").where(s"c2 > $lit"), Seq.empty)
+      checkAnswer(df.select("c1").where(s"c2 >= $lit"), Seq.empty)
+      checkAnswer(df.select("c1").where(s"c2 == $lit"), Seq.empty)
+      checkAnswer(df.select("c1").where(s"c2 <=> $lit"), Seq.empty)
+      checkAnswer(df.select("c1").where(s"c2 != $lit"), Row(1) :: Row(2) :: Nil)
+      checkAnswer(df.select("c1").where(s"c2 <= $lit"), Row(1) :: Row(2) :: Nil)
+      checkAnswer(df.select("c1").where(s"c2 < $lit"), Row(1) :: Row(2) :: Nil)
+    }
+  }
+
+  test("cases when literal is min") {
+    val t = "test_table"
+    withTable(t) {
+      Seq[(Integer, java.lang.Short, java.lang.Float)](
+        (1, 100.toShort, 3.14.toFloat), (2, Short.MinValue, Float.NegativeInfinity),
+        (3, null, null))
+        .toDF("c1", "c2", "c3").write.saveAsTable(t)
+      val df = spark.table(t)
+
+      var lit = Short.MinValue.toInt
+      checkAnswer(df.select("c1").where(s"c2 > $lit"), Row(1))
```

Comment:
can we put the `where` before `select`? It's weird to see we select only `c1` and then filter on `c2`.